README.markdown
# Paymo.biz Time Tracker

Paymo.biz Time Tracker is a Firefox add-on plugin that helps with tracking your time using the [paymo.biz](http://www.paymo.biz/) API.

Before using this plugin, you'll need to set your API key, username and password in plugin's Options.

This plugin adds a small statusbar icon. You can track your time using the popup menus you can access by left/right clicking on it.

You can install the add-on from the Mozilla Add-ons: [Paymo Time Tracker](https://addons.mozilla.org/en-US/firefox/addon/paymo-time-tracker/)


CHANGELOG
CHANGELOG

* 1.0 - 2010-09-02: first release
* 1.1 - 2010-09-03: multiple code changes required for Mozilla Add-ons approval
* 1.2 - 2010-09-05: NEW: customizable timesheet description, customizable statusbar text, disabling of icon animation
* 1.3 - 2010-09-07: NEW: auto-login option, status messages shown near icon for a few seconds after various commands
* 1.3.1 - 2010-09-08: FIX: status messages
* 1.3.4 - 2010-10-19: code changes required for Mozilla Add-ons approval
* 1.3.5 - 2010-12-21: API requests made with HTTPS
* 1.4.0 - 2011-02-16: NEW: pause time tracking, remember last task, fast starting/stoping of time tracking / FIXED: time description
* 1.4.1 - 2011-04-02: FIX: password not saved on Options dialog on Mac OS
* 1.4.2 - 2011-04-28: Enabled add-on for upcoming Firefox 5
* 1.4.3 - 2011-06-19: FIX: hide completed tasks
* 1.4.4 - 2011-09-05: Increased add-on compatiblity up to Firefox 8
* 1.4.5 - 2012-04-02: FIX: sometimes onLoadAfter() did not fire
* 1.4.6 - 2012-04-10: Code changes required for Mozilla Add-ons approval

chrome/content/date.js
/*
 * Date Format 1.2.3
 * Original (c) 2007-2009 Steven Levithan <stevenlevithan.com>
 * MIT license
 *
 * Includes enhancements by Scott Trenda <scott.trenda.net>
 * and Kris Kowal <cixar.com/~kris.kowal/>
 */

var paymottDate =
{
	// Some common format strings
	masks: {
		"default":      "ddd mmm dd yyyy HH:MM:ss",
		shortDate:      "m/d/yy",
		mediumDate:     "mmm d, yyyy",
		longDate:       "mmmm d, yyyy",
		fullDate:       "dddd, mmmm d, yyyy",
		shortTime:      "h:MM TT",
		mediumTime:     "h:MM:ss TT",
		longTime:       "h:MM:ss TT Z",
		isoDate:        "yyyy-mm-dd",
		isoTime:        "HH:MM:ss",
		isoDateTime:    "yyyy-mm-dd'T'HH:MM:ss",
		isoUtcDateTime: "UTC:yyyy-mm-dd'T'HH:MM:ss'Z'"
	},

	// Internationalization strings
	i18n: {
		dayNames: [
			"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat",
			"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"
		],
		monthNames: [
			"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
			"January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"
		]
	},
	
	format: function(date, mask, utc)
	{
		var	token = /d{1,4}|m{1,4}|yy(?:yy)?|([HhMsTt])\1?|[LloSZ]|"[^"]*"|'[^']*'/g,
			timezone = /\b(?:[PMCEA][SDP]T|(?:Pacific|Mountain|Central|Eastern|Atlantic) (?:Standard|Daylight|Prevailing) Time|(?:GMT|UTC)(?:[-+]\d{4})?)\b/g,
			timezoneClip = /[^-+\dA-Z]/g,
			pad = function (val, len) {
				val = String(val);
				len = len || 2;
				while (val.length < len) val = "0" + val;
				return val;
			};

		// Passing date through Date applies Date.parse, if necessary
		date = date ? new Date(date) : new Date;
		if (isNaN(date)) throw SyntaxError("invalid date");

		mask = String(this.masks[mask] || mask || this.masks["default"]);

		// Allow setting the utc argument via the mask
		if (mask.slice(0, 4) == "UTC:") {
			mask = mask.slice(4);
			utc = true;
		}

		var	_ = utc ? "getUTC" : "get",
			d = date[_ + "Date"](),
			D = date[_ + "Day"](),
			m = date[_ + "Month"](),
			y = date[_ + "FullYear"](),
			H = date[_ + "Hours"](),
			M = date[_ + "Minutes"](),
			s = date[_ + "Seconds"](),
			L = date[_ + "Milliseconds"](),
			o = utc ? 0 : date.getTimezoneOffset(),
			flags = {
				d:    d,
				dd:   pad(d),
				ddd:  this.i18n.dayNames[D],
				dddd: this.i18n.dayNames[D + 7],
				m:    m + 1,
				mm:   pad(m + 1),
				mmm:  this.i18n.monthNames[m],
				mmmm: this.i18n.monthNames[m + 12],
				yy:   String(y).slice(2),
				yyyy: y,
				h:    H % 12 || 12,
				hh:   pad(H % 12 || 12),
				H:    H,
				HH:   pad(H),
				M:    M,
				MM:   pad(M),
				s:    s,
				ss:   pad(s),
				l:    pad(L, 3),
				L:    pad(L > 99 ? Math.round(L / 10) : L),
				t:    H < 12 ? "a"  : "p",
				tt:   H < 12 ? "am" : "pm",
				T:    H < 12 ? "A"  : "P",
				TT:   H < 12 ? "AM" : "PM",
				Z:    utc ? "UTC" : (String(date).match(timezone) || [""]).pop().replace(timezoneClip, ""),
				o:    (o > 0 ? "-" : "+") + pad(Math.floor(Math.abs(o) / 60) * 100 + Math.abs(o) % 60, 4),
				S:    ["th", "st", "nd", "rd"][d % 10 > 3 ? 0 : (d % 100 - d % 10 != 10) * d % 10]
			};

		return mask.replace(token, function ($0) {
			return $0 in flags ? flags[$0] : $0.slice(1, $0.length - 1);
		});
	},

	diffSeconds: function(start, end)
	{
		var ts = start.getTime(), te = end.getTime(), diff = Math.floor((te - ts) / 1000);
		return diff;
	}
};


chrome/content/utils.js
/*
	The contents of this file are subject to the Mozilla Public License
	Version 1.1 (the "License"); you may not use this file except in
	compliance with the License. You may obtain a copy of the License at
	http://www.mozilla.org/MPL/

	Software distributed under the License is distributed on an "AS IS"
	basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
	License for the specific language governing rights and limitations
	under the License.

	The Original Code is Paymo.biz Time Tracker.

	The Initial Developer of the Original Code is Valentin Agachi, http://agachi.name/

	Copyright (C) 2010-2011, Valentin Agachi. All Rights Reserved.
*/

var paymottUtils =
{
	sConsole: null,
	debug: false,
	instant: null,

	onLoad: function()
	{
		this.sConsole = Components.classes["@mozilla.org/consoleservice;1"].getService(Components.interfaces.nsIConsoleService);

		var prefs = Components.classes["@mozilla.org/preferences-service;1"].getService(Components.interfaces.nsIPrefService);

		this.debug = prefs.getBranch("extensions.paymott.").getBoolPref("debug");
		this.instant = prefs.getBranch('browser.preferences.').getBoolPref('instantApply');
	},

	log: function(msg)
	{
		if (!this.sConsole || !this.debug)
			return;

		if ((msg.constructor != String))
			msg = JSON.stringify(msg);

		this.sConsole.logStringMessage('PaymoTT: ' + msg);
	}
};

window.addEventListener("load", function(){ paymottUtils.onLoad() }, false);


chrome/content/options.js
/*
	The contents of this file are subject to the Mozilla Public License
	Version 1.1 (the "License"); you may not use this file except in
	compliance with the License. You may obtain a copy of the License at
	http://www.mozilla.org/MPL/

	Software distributed under the License is distributed on an "AS IS"
	basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
	License for the specific language governing rights and limitations
	under the License.

	The Original Code is Paymo.biz Time Tracker.

	The Initial Developer of the Original Code is Valentin Agachi, http://agachi.name/

	Copyright (C) 2010-2011, Valentin Agachi. All Rights Reserved.
*/

var paymottOptions =
{

	onLoad: function()
	{
		paymottUtils.log('Options.onLoad()');

		paymottPassword.init();

		var username = this.$('username-input').value;

		var password = paymottPassword.get(username);
		this.$('password-input').value = password;
	},

	onDialogAccept: function(a)
	{
		paymottUtils.log('Options.onDialogAccept()');

		this.save();

		return true;
	},

	onUnload: function(a)
	{
		paymottUtils.log('Options.onUnload()');

		if (paymottUtils.instant)
			this.save();

		return true;
	},

	save: function()
	{
		var username = this.$('username-input').value,
			password = this.$('password-input').value;

		paymottPassword.add(username, password);
	},

	$: function(id){
		return document.getElementById(id);
	}
};

window.addEventListener("load", function(){ paymottOptions.onLoad() }, false);


chrome/content/password.js
/*
	The contents of this file are subject to the Mozilla Public License
	Version 1.1 (the "License"); you may not use this file except in
	compliance with the License. You may obtain a copy of the License at
	http://www.mozilla.org/MPL/

	Software distributed under the License is distributed on an "AS IS"
	basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
	License for the specific language governing rights and limitations
	under the License.

	The Original Code is Paymo.biz Time Tracker.

	The Initial Developer of the Original Code is Valentin Agachi, http://agachi.name/

	Copyright (C) 2010-2011, Valentin Agachi. All Rights Reserved.
*/

var paymottPassword = 
{
	hostname: 'chrome://paymott',
	formURL: null,
	httpRealm: 'paymo.biz',

	passManager: null,
	loginInfo: null,

	init: function()
	{
		this.passManager = Components.classes["@mozilla.org/login-manager;1"].getService(Components.interfaces.nsILoginManager);  
		this.loginInfo = new Components.Constructor("@mozilla.org/login-manager/loginInfo;1", Components.interfaces.nsILoginInfo, "init");
	},

	add: function(username, password)
	{
		this.log('Password.add(' + username + ', ********)');
		var loginOld= this.getLogin(username);
		var loginNew = new this.loginInfo(this.hostname, this.formURL, this.httpRealm, username, password, '', ''); 		
		if (loginOld)
		{
			if (loginOld.password != loginNew.password)
			{
				this.passManager.modifyLogin(loginOld, loginNew); 
			}
		}
		else
		{
			this.passManager.addLogin(loginNew);
		}
	},

	getLogin: function(username)
	{
		this.log('Password.getLogin(' + username + ')');
		try
		{
			var logins = this.passManager.findLogins({}, this.hostname, this.formURL, this.httpRealm);  
			for (var i = 0; i < logins.length; i++)
			{
				if (logins[i].username == username)
				{
					return logins[i];
				}
			}
		}
		catch (e) {
		}
		return null;
	},

	get: function(username)
	{
		this.log('Password.get(' + username + ')');
		var login = this.getLogin(username);
		if (login)
			return login.password;
		return '';
	},

	log: function(msg)
	{
		paymottUtils.log(msg);
	}
};




chrome/content/paymo-api.js
/*
	The contents of this file are subject to the Mozilla Public License
	Version 1.1 (the "License"); you may not use this file except in
	compliance with the License. You may obtain a copy of the License at
	http://www.mozilla.org/MPL/

	Software distributed under the License is distributed on an "AS IS"
	basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
	License for the specific language governing rights and limitations
	under the License.

	The Original Code is Paymo.biz Time Tracker.

	The Initial Developer of the Original Code is Valentin Agachi, http://agachi.name/

	Copyright (C) 2010-2011, Valentin Agachi. All Rights Reserved.
*/

var paymoAPI =
{
	apikey: null,
	authToken: null,
	authenticated: false,
	requests: [],


	login: function(username, password)
	{
		this._method('auth.login', { 'username': username, 'password': password });
	},

	logout: function()
	{
		this._method('auth.logout');
	},

	getProjects: function()
	{
		this._method('projects.getList', { 'include_tasks': 1 });
	},

	timeAdd: function(start, end, description)
	{
		var ts = paymottDate.format(start, 'yyyy-mm-dd HH:MM:ss');
		var te = paymottDate.format(end, 'yyyy-mm-dd HH:MM:ss');

		this._method('entries.add', {
			'start': ts,
			'end': te,
			'task_id': paymott.activeTask[0],
			'added_manually': 0,
			'description': description
		});
	},



	_method: function(method, args)
	{
		var url = 'https://api.paymo.biz/service/paymo.' + method + '?api_key=' + this.apikey + '&format=json';
		if (this.authenticated && this.authToken.length)
		{
			url += '&auth_token=' + this.authToken;
		}
		for (var k in args)
		{
			url += '&' + k + '=' + encodeURI(args[k]);
		}

		this.requests[this.requests.length] = [method, url];

		paymott.log(this.requests);

		if (this.authenticated || (method == 'auth.login'))
		{
			this._request();
		}
		else
		{
			return false;
		}
	},

	_complete: function(responseText)
	{
		if (!this.requests.length)
			return false;

		var request = this.requests.shift();

		this.log(request);
		this.log(responseText);

		var response = JSON.parse(responseText);

		if (response['status'] != 'ok')
		{
			this.log('_complete() error in response: ' + responseText);

			return false;
		}

		switch (request[0])
		{
		case 'auth.login':

			this.authToken = response['token']['_content'];
			this.authenticated = true;

			paymott.doLogin();

			break;		

		case 'auth.logout':

			this.authenticated = false;

			paymott.doLogout();

			break;		

		case 'projects.getList':

			var projects = [];
			for (var i = 0; i < response['projects']['project'].length; i++)
			{
				var project = response['projects']['project'][i];
				if (project['retired'] == 0)
				{
					projects.push(project);
				}
			}

			paymott.doProjects(projects);

			break;
		
		case 'entries.add':

			paymott.doTimeAdd();

			break;
		}
	},
	
	_request: function()
	{
		if (!this.requests.length)
			return false;

		var url = this.requests[0][1], method = 'GET';

		if (this.requests[0][0] == 'entries.add')
			method = 'POST';

		try
		{
			var x = new XMLHttpRequest();
			x.overrideMimeType("text/json");
			x.open(method, url, true);
			x.onreadystatechange = function()
			{
				if (x.status != 200)
					return;
			}
			x.onload = function()
			{
				var r = x.responseText;
				paymoAPI._complete(r);
			}
			x.send(null);
		}
		catch (e)
		{
			paymott.log('_request() exception');
			paymott.log(e);
		}

		if (this.requests.length > 1)
		{
			var timer = Components.classes["@mozilla.org/timer;1"].createInstance(Components.interfaces.nsITimer);  
			timer.initWithCallback(paymoAPITimer, 30000, Components.interfaces.nsITimer.TYPE_ONE_SHOT);
		}
	},
	
	log: function(msg)
	{
		paymottUtils.log(msg);
	}
};


var paymoAPITimer =
{
	notify: function(timer){
		paymoAPI._request();
	},
};


chrome/content/overlay.js
/*
	The contents of this file are subject to the Mozilla Public License
	Version 1.1 (the "License"); you may not use this file except in
	compliance with the License. You may obtain a copy of the License at
	http://www.mozilla.org/MPL/

	Software distributed under the License is distributed on an "AS IS"
	basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
	License for the specific language governing rights and limitations
	under the License.

	The Original Code is Paymo.biz Time Tracker.

	The Initial Developer of the Original Code is Valentin Agachi, http://agachi.name/

	Copyright (C) 2010-2011, Valentin Agachi. All Rights Reserved.
*/

var paymott =
{
	version: '@VERSION@',
	initialized: false,
	prefsBranch: null,
	preferences: {},
	statusPanel: null,
	popupProjects: null,

	miStop: null,
	miStart: null,

	ttDefault: null,
	ttUser: null,
	ttUsername: null,
	ttTask: null,
	ttTime: null,
	ttDiff: null,

	activeTask: null,
	active: false,
	paused: false,
	timeStart: 0,
	timePaused: 0,
	secondsPaused: 0,
	panelText: '',

	iconIndex: 1,

	timerTime: null,
	timerStatus: null,
	timerInit: null,

	onLoad: function()
	{
		this.strings = document.getElementById("paymott-strings");
	},

	onUnload: function()
	{
		if (this.active)
			this.onClickStop();
	},

	onLoadAfter: function()
	{
		if (this.initialized)
			return;

		this.timerTime = Components.classes["@mozilla.org/timer;1"].createInstance(Components.interfaces.nsITimer);
		this.timerStatus = Components.classes["@mozilla.org/timer;1"].createInstance(Components.interfaces.nsITimer);

		this.statusPanel = this.$('paymott-statuspanel');

		this.popupProjects = this.$('paymott-menupopup-projects');

		this.miStop = this.$('paymott-menuitem-stop');
		this.miPause = this.$('paymott-menuitem-pause');
		this.miStart = this.$('paymott-menuitem-start');

		this.$('paymott-tooltip-default').value += ' ver.' + this.version;

		this.ttUser = this.$('paymott-tooltip-user');
		this.ttUsername = this.$('paymott-tooltip-username');
		this.ttTask = this.$('paymott-tooltip-task');
		this.ttTime = this.$('paymott-tooltip-time');
		this.ttDiff = this.$('paymott-tooltip-diff');

		paymottPreferencesObserver.register();

		this.prefsBranch = Components.classes["@mozilla.org/preferences-service;1"].getService(Components.interfaces.nsIPrefService).getBranch("extensions.paymott.");

		this.loadPreferences();

		if (this.preferences.fastStart)
		{
			this.statusPanel.setAttribute('popup', '');
			this.statusPanel.addEventListener('click', function(ev){
				paymott.onPanelClick(ev);
			}, false);
		}

		this.initialized = true;

		this.log('onLoadAfter() complete');

		paymottPassword.init();

		if (this.preferences.autoLogin)
		{
			this.onClickLogin();
		}

		// TODO load dynamically paymo-api.js, date.js
	},

	loadPreferences: function()
	{
		this.log('loadPreferences()');

		this.preferences = {
			apikey: this.prefsBranch.getCharPref("apikey"),
			username: this.prefsBranch.getCharPref("username"),
			autoLogin: this.prefsBranch.getBoolPref("autoLogin"),
			panelShow: this.prefsBranch.getBoolPref("panelShow"),
			panelFormat: this.prefsBranch.getCharPref("panelFormat"),
			iconAnimate: this.prefsBranch.getBoolPref("iconAnimate"),
			rememberTask: this.prefsBranch.getBoolPref("rememberTask"),
			fastStart: this.prefsBranch.getBoolPref("fastStart"),
			timeDescription: this.prefsBranch.getCharPref("timeDescription"),
			lastTask: this.prefsBranch.getIntPref("lastTask")
		};

		paymoAPI.apikey = this.preferences.apikey;

		if (!this.preferences.apikey.length)
		{
			this.ttTask.value = 'Please setup the API key and login data in Options!';
		}
		this.ttTask.hidden = (this.preferences.apikey.length > 0);

		return true;
	},

	savePreferences: function(prefs)
	{
		for (var k in prefs)
		{
			if (typeof(prefs[k]) == 'string')
				this.prefsBranch.setCharPref(k, prefs[k]);

			if (typeof(prefs[k]) == 'number')
				this.prefsBranch.setIntPref(k, prefs[k]);

			if (typeof(prefs[k]) == 'boolean')
				this.prefsBranch.setBoolPref(k, prefs[k]);
		}
	},



	onClickOptions: function()
	{
		window.openDialog("chrome://paymott/content/options.xul", "", null)
	},

	onClickLogin: function()
	{
		this.log('onClickLogin()');

		var password = paymottPassword.get(this.preferences.username);

		this.showStatus('Logging in ...');

		paymoAPI.login(this.preferences.username, password);
	},

	doLogin: function()
	{
		this.log('doLogin()');

		this.statusPanel.contextMenu = 'paymott-menupopup-projects';
		this.statusPanel.image = 'chrome://paymott/skin/clock-inactive.png';

		this.$('paymott-menuitem-loggedin').label = 'Logged in as: ' + this.preferences.username;

		this.ttUser.hidden = false;
		this.ttUsername.textContent = this.preferences.username;

		this.showStatus('Logged in as ' + this.preferences.username);

		paymoAPI.getProjects();
	},

	onClickLogout: function()
	{
		this.log('onClickLogout()');

		if (this.active)
			this.onClickStop();

		paymoAPI.logout();

		this.statusPanel.contextMenu = 'paymott-menupopup-logout';
		this.statusPanel.image = 'chrome://paymott/skin/paymo-bw-16x16.png';

		this.ttUser.hidden = true;
		this.ttTask.hidden = true;

		this.miStop.disabled = true;
		this.miPause.disabled = true;
		this.miStart.disabled = true;
	},

	doLogout: function()
	{
		this.panelText = '';
		this.showStatus('Logged out');
	},

	onClickRefresh: function()
	{
		paymoAPI.getProjects();
	},

	doProjects: function(projects)
	{
		this.log('doProjects() - lastTask: ' + this.preferences.lastTask);

		this.activeTask = null;

		var items = this.popupProjects.children;
		if (items.length > 6)
		{
			for (var i = items.length - 1; i > 5; i--)
			{
				this.popupProjects.removeChild(items[i]);
			}
		}

		for (var i = 0; i < projects.length; i++)
		{
			var project = projects[i];

			var lists = project['task_lists']['task_list'];

			for (var j = 0; j < lists.length; j++)
			{
				var list = lists[j];

				if (list['tasks']['task'])
				{
					var tasks = list['tasks']['task'];

					for (var k = 0; k < tasks.length; k++)
					{
						var task = tasks[k];

						if (task['complete'] == 1) continue;

						var child = document.createElement('menuitem');
						child.setAttribute('label', project['name'] + ': ' + task['name']);
						child.setAttribute('type', 'radio');
						child.setAttribute('id', 'task-' + task['id']);
						child.addEventListener('click', function(e){ paymott.onClickTaskItem(e) }, false);
						this.popupProjects.appendChild(child);

						if (this.preferences.rememberTask && (task['id'] == this.preferences.lastTask))
						{
							child.setAttribute('checked', true);
							child.click();
						}
					}
				}
			}
		}

		this.ttTask.hidden = true;
	},

	onClickTaskItem: function(event)
	{
		this.activeTask = [parseInt(event.target.id.split('-')[1]), event.target.getAttribute('label')];

		this.ttTask.hidden = false;

		this.miStop.disabled = true;
		this.miPause.disabled = true;
		this.miStart.disabled = false;

		this.showStatus('Active task: ' + this.activeTask[1]);

		this.savePreferences({ lastTask: this.activeTask[0] });

		if (this.preferences.panelShow && this.preferences.panelFormat.match(/%p/))
		{
			this.panelText = this.getPanelText('%p');
		}
	},

	onClickStart: function()
	{
		this.active = true;

		this.timeStart = new Date();
		this.secondsPaused = 0;

		this.statusPanel.image = 'chrome://paymott/skin/clock-active-1.png';
		if (this.preferences.panelShow)
		{
			this.showPanelText(this.getPanelText());
		}

		this.miStop.disabled = false;
		this.miPause.disabled = false;
		this.miStart.disabled = true;

		this.iconIndex = 1;
		this.timerTime.init(paymottTimer, 1000, Ci.nsITimer.TYPE_REPEATING_PRECISE);
	},

	onClickPause: function(stop)
	{
		this.paused = !this.paused;

		if (this.paused)
		{
			this.timePaused = new Date();

			this.miPause.label = 'Resume';

			this.statusPanel.image = 'chrome://paymott/skin/clock-paused.png';

			this.timerTime.cancel();

			this.onTimerIcon();
		}
		else
		{
			this.secondsPaused += paymottDate.diffSeconds(this.timePaused, new Date());

			this.miPause.label = 'Pause';

			if (!stop)
			{
				this.statusPanel.image = 'chrome://paymott/skin/clock-active-1.png';

				this.timerTime.init(paymottTimer, 1000, Ci.nsITimer.TYPE_REPEATING_PRECISE);
			}
		}
	},

	onClickStop: function()
	{
		if (this.paused)
			this.onClickPause(true);

		this.active = false;

		this.showStatus('Recording time ...');

		this.timeEnd = new Date();
		this.timeEnd = new Date(this.timeEnd.getTime() - this.secondsPaused * 1000);

		paymoAPI.timeAdd(this.timeStart, this.timeEnd, this.preferences.timeDescription);

		this.statusPanel.image = 'chrome://paymott/skin/clock-inactive.png';

		this.miStop.disabled = true;
		this.miPause.disabled = true;
		this.miStart.disabled = false;

		this.timerTime.cancel();

		if (this.preferences.panelShow && this.preferences.panelFormat.match(/%p/))
		{
			this.panelText = this.getPanelText('%p');
		}
	},


	onPanelClick: function(event)
	{
		if (!this.initialized)
			this.onLoadAfter();

		if (event.button != 0)
			return;

		if (this.activeTask == null)
			return;

		this.active ? this.onClickStop() : this.onClickStart();
	},


	doTimeAdd: function()
	{
		var diff = this.timeDiff(this.timeStart, this.timeEnd, 0, true);
		this.showStatus('Time recorded: ' + diff);
	},

	onTimerIcon: function()
	{
		if (this.preferences.iconAnimate && !this.paused)
		{
			this.iconIndex++;
			if (this.iconIndex > 8)
				this.iconIndex = 1;
			this.statusPanel.image = 'chrome://paymott/skin/clock-active-' + paymott.iconIndex + '.png';
		}

		if (this.preferences.panelShow)
		{
			this.panelText = this.getPanelText();
			this.showPanelText(this.panelText);
		}
	},

	getPanelText: function(f)
	{
		var t = (f ? f : this.preferences.panelFormat);

		t = t.replace('%p', this.activeTask[1]);

		if (this.active)
		{
			if (!this.paused)
			{
				var diff = this.timeDiff(this.timeStart, new Date(), this.secondsPaused - 1, true);
				t = t.replace('%s', diff);
				t = t.replace('%t', diff.substr(0, 5));
			}
			else
			{
				t = t.replace('%s', 'PAUSED');
				t = t.replace('%t', 'PAUSED');
			}
		}

		return t;
	},

	onTooltipShow: function()
	{
		if (!this.initialized)
			this.onLoadAfter();

		this.ttTime.hidden = !this.active;

		if (this.activeTask)
		{
			this.ttTask.value = this.activeTask[1];
		}
		if (this.active)
		{
			var secPaused = this.secondsPaused;

			if (this.paused)
				secPaused += paymottDate.diffSeconds(this.timePaused, new Date());

			this.ttDiff.textContent = this.timeDiff(this.timeStart, new Date(), secPaused, true);
		}
	},

	showStatus: function(msg)
	{
		if (!this.initialized)
			this.onLoadAfter();

		this.showPanelText(msg);

		this.timerStatus.cancel();

		var _event = {
			notify: function(){
				paymott.clearPanelText();
			}
		};

		this.timerStatus.initWithCallback(_event, 3000, Ci.nsITimer.TYPE_ONE_SHOT);
	},

	showPanelText: function(msg)
	{
		if (!this.initialized)
			this.onLoadAfter();

		this.statusPanel.className = 'statusbarpanel-iconic-text';
		this.statusPanel.label = msg;
	},

	clearPanelText: function()
	{
		if (!this.initialized)
			this.onLoadAfter();

		if (!this.panelText.length)
		{
			this.statusPanel.className = 'statusbarpanel-iconic';
		}
		this.statusPanel.label = this.panelText;
	},

	log: function(msg)
	{
		paymottUtils.log(msg);
	},

	timeDiff: function(start, end, paused, seconds)
	{
		var diff = paymottDate.diffSeconds(start, end) - paused;

		var h = Math.floor(diff / 3600);
		diff = diff % 3600;
		var m = Math.floor(diff / 60);
		var s = diff % 60;

		var t = (h.toString().length == 1 ? '0' + h : h);
		t += ':' + (m.toString().length == 1 ? '0' + m : m);
		if (seconds)
		{
			t += ':' + (s.toString().length == 1 ? '0' + s : s);
		}

		return t;
	},

	$: function(id){
		return document.getElementById(id);
	}
};



var paymottTimer =
{
	notify: function(){
		paymott.onLoadAfter();
	},
	observe: function(){
		paymott.onTimerIcon();
	}
};



var paymottPreferencesObserver =
{
	register: function(){
		var prefService = Cc["@mozilla.org/preferences-service;1"].getService(Ci.nsIPrefService);
		this._branch = prefService.getBranch("extensions.paymott.");
		this._branch.QueryInterface(Ci.nsIPrefBranch2);
		this._branch.addObserver("", this, false);
	},

	unregister: function(){
		if(!this._branch) {
			return;
		}
		this._branch.removeObserver("", this);
	},

	observe: function(aSubject, aTopic, aData) {
		if (aTopic != "nsPref:changed") {
			return;
		}
		paymott.loadPreferences();
	}
};



// this doesn't always trigger on some configurations
//window.addEventListener("load", function(){ paymott.onLoad() }, false);

// onLoad workaround - fire after 10 seconds
paymott.timerInit = Components.classes["@mozilla.org/timer;1"].createInstance(Components.interfaces.nsITimer);
paymott.timerInit.initWithCallback(paymottTimer, 10000, Ci.nsITimer.TYPE_ONE_SHOT);


window.addEventListener("unload", function(){ paymott.onUnload() }, false);


